script-runner/AutomatedTestRunner.hs

```diff
@@ -0,0 +1,400 @@
+      liftIO $ writeBChan eventChan $ CESlotStart $ SlotStart (getEpochIndex $ siEpoch slotid) (getSlotIndex $ siSlot slotid)
+      case Map.lookup slotid (slotTriggers realScript) of
+        Just (SlotTrigger act) -> runAction act
+        Nothing -> pure ()
+      pure ()
+    realScript = getScript script
+    errhandler :: Show e => e -> AuxxMode ()
+    errhandler e = print e
+    runAction :: (Dict HasConfigurations -> Diffusion AuxxMode -> AuxxMode ()) -> AuxxMode ()
+    runAction act = do
+      act Dict diffusion `catch` errhandler @SomeException
+    realWorker = do
+      mapM_ (\(SlotTrigger act) -> runAction act) (startupActions realScript)
+      onNewSlot (configEpochSlots genesisConfig) defaultOnNewSlotParams handler
+      pure ()
+  realWorker `catch` errhandler @SomeException
+
+data TestScript a => InputParams a = InputParams
+  { ipEventChan :: BChan CustomEvent
+  , ipReplyChan :: BChan Reply
+  , ipScriptGetter :: HasEpochSlots => IO a
+  }
+data TestScript a => InputParams2 a = InputParams2
+  { ip2EventChan :: BChan CustomEvent
+  , ip2ReplyChan :: BChan Reply
+  , ip2Script :: a
+  }
+
+runScript :: TestScript a => (HasEpochSlots => IO a) -> IO ()
+runScript scriptGetter = withCompileInfo $ do
+  (eventChan, replyChan, asyncUi) <- runUI
+  opts <- getScriptRunnerOptions
+  let
+    inputParams = InputParams eventChan replyChan scriptGetter
+    loggingParams = CLI.loggingParams loggerName (srCommonNodeArgs opts)
+  loggerBracket "script-runner" loggingParams . logException "script-runner" $ do
+    thing opts inputParams
+    pure ()
+  liftIO $ writeBChan eventChan QuitEvent
+  finalState <- wait asyncUi
+  --print finalState
+  pure ()
+
+runUI :: IO (BChan CustomEvent, BChan Reply, Async AppState)
+runUI = do
+  eventChan <- newBChan 10
+  replyChan <- newBChan 10
+  let
+    app = App
+      { appDraw = ui
+      , appChooseCursor = showFirstCursor
+      , appHandleEvent = handleEvent
+      , appStartEvent = \x -> pure x
+      , appAttrMap = const $ attrMap defAttr []
+      }
+    state :: AppState
+    state = AppState 0 Nothing "" Nothing replyChan
+    go :: IO AppState
+    go = do
+      finalState <- customMain (mkVty defaultConfig) (Just eventChan) app state
+      writeBChan replyChan TriggerShutdown
+      pure finalState
+  brick <- async go
+  pure (eventChan, replyChan, brick)
+
+getGenesisConfig :: Example Config
+getGenesisConfig = do
+  oldsb <- get
+  pure $ sbGenesisConfig oldsb
+
+data SlotCreationFailure = SlotCreationFailure { msg :: Text, slotsInEpoch :: SlotCount } deriving Show
+instance Exception SlotCreationFailure where
+
+onStartup :: (Dict HasConfigurations -> Diffusion AuxxMode -> AuxxMode ()) -> Example ()
+onStartup action = do
+  oldsb <- get
+  let
+    oldscript = sbScript oldsb
+    script = oldscript {
+      startupActions = [ SlotTrigger action ] <> (startupActions oldscript)
+    }
+    newsb = oldsb {
+      sbScript = script
+    }
+  put newsb
+  pure ()
```

Comment:
The type of `put newsb` should probably be `Example ()` so the `pure ()` can be dropped.

Reply:
it already is, the `pure ()` was just a left-over, from when the function was entirely empty and ghc complained about an empty do block